tempesta_fw/str_avx2.S

```diff
@@ -0,0 +1,2296 @@
+#define __CUST_COOKIE_0		$__CUSTOM+0x180
+#define __CUST_COOKIE_1		$__CUSTOM+0x1a0
+
+.extern __tfw_lct
+
+.section .text
+
+#ifdef DEBUG
+.extern	tfw_dbg_vprint32
+
+/**
+ * Use the debug routine to print 32-byte vector residing by memory address
+ * residing in @R. The macro is inserted into a function body to print current
+ * function arguments, so we save and restore the registers context.
+ * Usage exmple (still good for 16 byte registers since we use 32 byte tables):
+ *	...
+ *	DPRN_V	%rcx
+	vlddqu	(%rcx), %xmm15
+ */
+.macro DPRN_V	R
+	push	%rax
+	push	%rdi
+	push	%rsi
+	push	%rdx
+	push	%rcx
+	push	%r8
+	push	%r9
+
+	movq	$dbg_prefix_vec, %rdi
+	movq	\R, %rsi
+	call	tfw_dbg_vprint32
+
+	pop	%r9
+	pop	%r8
+	pop	%rcx
+	pop	%rdx
+	pop	%rsi
+	pop	%rdi
+	pop	%rax
+.endm
+
+/**
+ * The same as the above, but prints any register value.
+ * Can be used to any size registers, but top bytes for small register will
+ * contain zeros from stack.
+ */
+.macro __DPRN_R	R MOV
+	push	%rsi
+	push	%rdi
+	subq	$64, %rsp
+	/*
+	 * Use 32-byte aligned address to avoid #GF.
+	 * Do not align RSP directly and do not modify RBP to avoid objtool
+	 * claims on stack alignment.
+	 */
+	lea	32(%rsp), %rsi
+	andq	$~0x1f, %rsi
+	/* Write zeroes to clear extra space for small registers. */
+	xor	%rdi, %rdi
+	movq	%rdi, (%rsi)
+	movq	%rdi, 0x8(%rsi)
+	movq	%rdi, 0x10(%rsi)
+	movq	%rdi, 0x18(%rsi)
+
+	\MOV	\R, (%rsi)
+	DPRN_V	%rsi
+
+	addq	$64, %rsp
+	pop	%rdi
+	pop	%rsi
+.endm
+
+.macro DPRN_VR	R
+	__DPRN_R \R vmovaps
+.endm
+
+.macro DPRN_R	R
+	/* Use RAX as temporary storage in R is either RSI or RDI. */
+	push	%rax
+	movq	\R, %rax
+
+	__DPRN_R %rax movq
+
+	pop %rax
+.endm
+#endif
+
+.macro __STRTOLOWER_SMALL_STR	SZ
+	movzbl	\SZ(%rsi), %eax
+	movzbl	__tfw_lct(%rax), %eax
+	movb	%al, \SZ(%rdi)
+.endm
+
+.macro __STRTOLOWER_TEST_LEN	SZ R1 R2 L
+	leal	\SZ(%rax), \R1
+	movslq	\R1, \R2
+	cmpq	\R2, %rdx
+	jnb	\L
+.endm
+
+/**
+ * Convert a string to lower case.
+ *
+ * The function uses byte by byte case conversion using __tfw_lct table and
+ * lookup table for the string length to process string suffix and strings
+ * smaller than 9 bytes. Next it uses 128, 64, 32, 16 and 8 byte blocks
+ * processing using vector instructions. The pseudocode for 32 byte block:
+ *
+ *	__m256i v = _mm256_lddqu_si256((void *)src);
+ *	__m256i sub = _mm256_sub_epi8(v, __A);
+ *	__m256i cmp_r = _mm256_cmpgt_epi8(__D, sub);
+ *	__m256i lc = _mm256_and_si256(cmp_r, __CASE);
+ *	__m256i r = _mm256_or_si256(v, lc);
+ *	_mm256_storeu_si256((__m256i *)dest, r);
+ */
+ENTRY(__tfw_strtolower_avx2)
+	leaq	8(%rsp), %r10
+	andq	$-32, %rsp
+	xorl	%eax, %eax
+	movabsq	$CASE, %rcx
+	pushq	-8(%r10)
+	pushq	%rbp
+	movq	%rsp, %rbp
+	pushq	%r12
+	pushq	%r10
+	pushq	%rbx
+	subq	$88, %rsp
+	movq	__A, %r9
+	movq	__D, %r8
+
+	/* Process 8 byte prefix of suffix of the string. */
+.str2low_prefix_suffix:
+	cmpq	$8, %rdx
+	ja	.str2low_test_len128
+	movq	.str2low_switch(,%rdx,8), %rax
+	jmpq	*%rax
```

Comment:
These indirect jumps make objtool generate a warning: "warning: objtool: __tfw_strtolower_avx2()+0x46: indirect jump found in RETPOLINE build". To suppress it we need either annotate these jumps like that:
```diff
diff --git a/tempesta_fw/str_avx2.S b/tempesta_fw/str_avx2.S
index 5f890fe6..df3e2e2f 100644
--- a/tempesta_fw/str_avx2.S
+++ b/tempesta_fw/str_avx2.S
@@ -24,6 +24,7 @@
 #include <linux/linkage.h>
 #include <asm/alternative-asm.h>
 #include <asm/export.h>
+#include <asm/nospec-branch.h>
 
 #define CASE	0x2020202020202020
 
@@ -476,6 +477,7 @@ ENTRY(__tfw_strtolower_avx2)
 	cmpq	$8, %rdx
 	ja	.str2low_test_len128
 	movq	.str2low_switch(,%rdx,8), %rax
+	ANNOTATE_RETPOLINE_SAFE
 	jmpq	*%rax
 .section .rodata
 .align	8
@@ -653,6 +655,7 @@ ENTRY(__tfw_stricmp_avx2)
 
 	/* Process short strings below 8 bytes in length. */
 	movq	.stricmp_switch(,%rdx,8), %rax
+	ANNOTATE_RETPOLINE_SAFE
 	jmpq	*%rax
 .section .rodata
 .align 8
@@ -1009,6 +1012,7 @@ ENTRY(__tfw_stricmp_avx2_2lc)
 	ja	.sic2lc_short
 
 	movq	.sic2lc_switch(,%rdx,8), %rax
+	ANNOTATE_RETPOLINE_SAFE
 	jmpq	*%rax
 .section	.rodata
 .align 8
```
or enable Spectre mitigation:
```diff
diff --git a/tempesta_fw/str_avx2.S b/tempesta_fw/str_avx2.S
index 5f890fe6..d5d2334c 100644
--- a/tempesta_fw/str_avx2.S
+++ b/tempesta_fw/str_avx2.S
@@ -24,6 +24,7 @@
 #include <linux/linkage.h>
 #include <asm/alternative-asm.h>
 #include <asm/export.h>
+#include <asm/nospec-branch.h>
 
 #define CASE	0x2020202020202020
 
@@ -476,7 +477,7 @@ ENTRY(__tfw_strtolower_avx2)
 	cmpq	$8, %rdx
 	ja	.str2low_test_len128
 	movq	.str2low_switch(,%rdx,8), %rax
-	jmpq	*%rax
+	JMP_NOSPEC	%rax
 .section .rodata
 .align	8
 .str2low_switch:
@@ -653,7 +654,7 @@ ENTRY(__tfw_stricmp_avx2)
 
 	/* Process short strings below 8 bytes in length. */
 	movq	.stricmp_switch(,%rdx,8), %rax
-	jmpq	*%rax
+	JMP_NOSPEC	%rax
 .section .rodata
 .align 8
 .stricmp_switch:
@@ -1009,7 +1010,7 @@ ENTRY(__tfw_stricmp_avx2_2lc)
 	ja	.sic2lc_short
 
 	movq	.sic2lc_switch(,%rdx,8), %rax
-	jmpq	*%rax
+	JMP_NOSPEC	%rax
 .section	.rodata
 .align 8
 .sic2lc_switch:
```

Reply:
We check jump array bounds against a constant, encoded in an instruction opcode, so there is no cache miss on boundary check [required for Spectre](https://spectreattack.com/spectre.pdf). `JPM_NOSPEC` gives about 15% performance impact in comparison with `ANNOTATE_RETPOLINE_SAFE`: 101ms vs 85ms in the `tfw_str` microbenchmark.